pyccl/halos/profiles.py

```diff
@@ -0,0 +1,824 @@
+            should be `f(cosmo, M, a, mdef)`, where `cosmo`
+            is a :obj:`Cosmology` object, `M` is a halo mass
+            in units of M_sun, `a` is the scale factor and
+            `mdef` is a :obj:`MassDef` object.
+        tilt (:obj:`function`): the power law index of the
+            profile. The signature of this function should
+            be `f(cosmo, a)`.
+    """
+    name = 'PowerLaw'
+
+    def __init__(self, r_scale, tilt):
+        self.r_s = r_scale
+        self.tilt = tilt
+        super(HaloProfilePowerLaw, self).__init__()
+
+    def _get_plaw_fourier(self, cosmo, a):
+        # This is the optimal value for a pure power law
+        # profile.
+        return self.tilt(cosmo, a)
+
+    def _get_plaw_projected(self, cosmo, a):
+        # This is the optimal value for a pure power law
+        # profile.
+        return -3 - self.tilt(cosmo, a)
+
+    def _real(self, cosmo, r, M, a, mass_def):
+        r_use = np.atleast_1d(r)
+        M_use = np.atleast_1d(M)
+
+        # Compute scale
+        rs = self.r_s(cosmo, M_use, a, mass_def)
+        tilt = self.tilt(cosmo, a)
+        # Form factor
+        prof = (r_use[None, :] / rs[:, None])**tilt
+
+        if np.ndim(r) == 0:
+            prof = np.squeeze(prof, axis=-1)
+        if np.ndim(M) == 0:
+            prof = np.squeeze(prof, axis=0)
+        return prof
+
+
+class HaloProfileNFW(HaloProfile):
+    """ Navarro-Frenk-White (astro-ph:astro-ph/9508025) profile.
+
+    .. math::
+       \\rho(r) = \\frac{\\rho_0}
+       {\\frac{r}{r_s}\\left(1+\\frac{r}{r_s}\\right)^2}
+
+    where :math:`r_s` is related to the spherical overdensity
+    halo radius :math:`R_\\Delta(M)` through the concentration
+    parameter :math:`c(M)` as
+
+    .. math::
+       R_\\Delta(M) = c(M)\\,r_s
+
+    and the normalization :math:`\\rho_0` is
+
+    .. math::
+       \\rho_0 = \\frac{M}{4\\pi\\,r_s^3\\,[\\log(1+c) - c/(1+c)]}
+
+    Args:
+        c_M_relation (:obj:`Concentration`): concentration-mass
+            relation to use with this profile.
+        fourier_analytic (bool): set to `True` if you want to compute
+            the Fourier profile analytically (and not through FFTLog).
+            Default: `False`.
+        projected_analytic (bool): set to `True` if you want to
+            compute the 2D projected profile analytically (and not
+            through FFTLog). Default: `False`.
+        cumul2d_analytic (bool): set to `True` if you want to
+            compute the 2D cumulative surface density analytically
+            (and not through FFTLog). Default: `False`.
+        truncated (bool): set to `True` if the profile should be
+            truncated at :math:`r = R_\\Delta` (i.e. zero at larger
+            radii.
+    """
+    name = 'NFW'
+
+    def __init__(self, c_M_relation,
+                 fourier_analytic=False,
+                 projected_analytic=False,
+                 cumul2d_analytic=False,
+                 truncated=True):
```

Comment:
Why these defaults, especially the one with `truncated=True`? Are the analytic computations less efficient?

Reply:
That's what I thought initially, but I've been doing a few test in a different branch and it doesn't seem to be the case. I've changed `projected_analytic` to `True`.
I've kept `projected_analytic` and `cumul2d_analytic` to `False`, since we don't have those for truncated profiles.
I think `truncated=True` makes more sense, since the halo mass is only well defined for truncated profiles (otherwise it's infinite).

Reply:
I'm worried the truncated=True default is too surprising.

Reply:
I could document this?

Reply:
Yeah or call the entire class `TruncatedNFWProfile` or whatver.

Reply:
Done!